### DISCUSSION TITLE
Is it neccessary to shuffle a block gradient kernel for a multitaskmultioutput model or can this operation be avoided?

### DISCUSSION THREAD

**Discussion by JGittles** [cornellius-gp/gpytorch#2079] [Q&A]
Hi All,

I have been deep in the weeds attempting to implement a keops version of the rbfwithgrad kernel (https://docs.gpytorch.ai/en/latest/_modules/gpytorch/kernels/rbf_kernel_grad.html) to be used as a drag and drop replacement to improve scale of this kernel.

Currently, I am about 99% of the way there and have hit a wall with how to do a perfect shuffle permutation on my kernel matrix within the confines of the keops library. I have opened up a ticket on their repo but also wanted to check here to see if I even *need* to perform this operation.

At the moment, my code is essentially a direct copy of the example here (https://docs.gpytorch.ai/en/stable/examples/08_Advanced_Usage/Simple_GP_Regression_Derivative_Information_2d.html) generalized to 3D.

Is there any way I can re-order my input values or change the structure of my model in order to use my kernel in the typical block format? I do not necessarily need to learn hyperparameters if that is a potential hangup for an alternative structure.
![rbf-block](https://user-images.githubusercontent.com/58196263/180843097-cb41574b-01e4-4475-8aad-0e12845b7891.png)

Thanks!

**Comment by gpleiss** [comment]
Our `MultitaskMultivariateNormal` distribution assumes that the data are ordered `num_data x num_tasks`. The covariance matrix should therefore be ordered in a similar way: `[k(x1, x1); dk(x1, x1); k(x2, x1); dk(x2, x1); ...]`. This is what the shuffle operation is doing.

The `MultitaskMultivariateNormal` distribution also has an `interleaved=True` option, which would then expect the data as `num_tasks x num_data`. The kernel would then be ordered as `[k(x1, x1); k(x2, x1); ..., k(xn, x1); dk(x1, x1); dk(x2, x1); ...]` - which wouldn't require the shuffling (as far as I can tell). However, you might have to change a lot of code (and maybe the internals of `MultitaskGaussianLikelihood`) since most of our code assumes the `MultitaskMultivariateNormal` distributions are not interleaved.

Good luck!
